### DISCUSSION TITLE
When making a gif background transparent, can't get -dispose to remove previous frames

### DISCUSSION THREAD

**Discussion by bertiebaggio** [ImageMagick/ImageMagick#5529] [Help] [answered]
Hi, I am trying to remove the white background from an animated hourglass:

[![hourglass](https://c.tenor.com/5JWmM_Hd3rIAAAAC/loading-windows98.gif)](https://c.tenor.com/5JWmM_Hd3rIAAAAC/loading-windows98.gif)

I have tried a few incantations of [dispose](https://imagemagick.org/script/command-line-options.php?#dispose), admittedly probably based on old IM6 examples.

    $ magick [-background White] -dispose [Previous|Background] loading-windows98.gif [-coalesce] -fuzz 5% -transparent White [-layers optimize] loading-windows98-transparent.gif

Most variants I have tried are in square brackets.

This produces an hourglass with transparent background but with previous frames present, most noticeable towards the end of the animation:

![loading-windows98-transparent3](https://user-images.githubusercontent.com/7524620/188266475-c001681e-0f96-49e3-a49a-a8890c6c9ff4.gif)

**Addendum**: weirdly, if I:
 1. cheat and use [an online converter](https://onlinegiftools.com/create-transparent-gif) to remove the background
 2. recolour the hourglass
 3. add a drop shadow as in [the old layer composition example](https://legacy.imagemagick.org/Usage/anim_mods/#composite)
 4. re-remove the gray background **without any dispose** (`convert composite_green_hourglass.gif -fuzz 15% -transparent Gray composite_green_hourglass-transparent.gif`)

I get a version that has the right behaviour: transparency and previous frames disposed of. 

![composite_green_hourglass-transparent](https://user-images.githubusercontent.com/7524620/188266903-c38839d2-5d20-4f76-a0d3-2ddec9443372.gif)
_(I know it looks kinda terrible due to 1-bit alpha in gifs, I just wanted to see how close I could get it)_

What is the online converter doing different?

I must have a gap in my knowledge, and would appreciate pointers in the right direction. Thanks in advance :)

---
```
$ magick -version
Version: ImageMagick 7.1.0-47 Q16-HDRI x86_64 20393 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenCL OpenMP(4.5) 
Delegates (built-in): bzlib cairo djvu fontconfig freetype heic jbig jng jp2 jpeg jxl lcms lqr ltdl lzma openexr pangocairo png raqm raw rsvg tiff webp wmf x xml zip zlib
Compiler: gcc (12.2)
```

**Comment by snibgo** [comment, marked as answer]
> Hi, I am trying to remove the white background ...

As you realise, none of the input pixels are white, so you need "`-fuzz`".

How about this (Windows BAT syntax).
```
magick ^
  hourglass.gif ^
  -layers Coalesce ^
  -fuzz 5%% ^
  -transparent White ^
  -set dispose Background ^
  -layers optimize ^
  hg2.gif
```
The "`-layers Coalesce`" is not needed for your input.

**Reply by bertiebaggio** [reply]
Thanks for taking the time to answer snigbo; you're right the background is off-white so I had `-fuzz` set! I did try `-set dispose Background` (and `Previous`) as well as the -dispose pre-input option, but...

    $ magick loading-windows98.gif -coalesce -fuzz 5% -transparent White -set dispose Background -layers optimize loading-windows98-transparent4.gif

this indeed works!

**It seems that -set dispose Background is more sensitive to where it is in the argument list than I previously appreciated**. Is this, I am wildly guessing, so that disposal is done after the background is transparent-ed out? I'd love to hear a slightly more technical reason / more coherent explanation :)

**Comment by snibgo** [comment]
In IM v7, each option is executed in the order it appears in the command. "`-set dispose Background`" changes metadata for each frame in memory, so it doesn't matter when it is done, provided it is *after* the input has been read, and *before* "`-layers optimize`".

This also works:
```
magick ^
  hourglass.gif ^
  -layers Coalesce ^
  -set dispose Background ^
  -fuzz 5%% ^
  -transparent White ^
  -layers optimize ^
  hg3.gif
```
On "`-dispose`", see https://imagemagick.org/Usage/anim_basics/#dispose. Note that "`-dispose`" is a setting that will affect any frames that are read after that option, and "`-set dispose`" is an operation that affects all frames (changing their metadata) that are currently in memory.

"`-dispose`" before reading a GIF is a bit useless, as it will be overridden by whatever is in the GIF file.

"`-set dispose`" changes metadata, not pixels, so it can come before or after making white transparent.

On the difference between a *setting* and an *operation*, see https://imagemagick.org/script/command-line-processing.php#option

Yeah, this is non-intuitive. "`-dispose`" (with no "set") is a setting, but "`-set dispose`" (with "set") is an operation. Such is life.
